chanfitness/chanfitness.go

```diff
@@ -0,0 +1,300 @@
+// Note that this does not equal channel lifetime because data is not currently
+// persisted.
+//
+// Uptime: the total time within the channel's lifetime that its peer has been
+// observed as being online.
+package chanfitness
+
+import (
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/channelnotifier"
+	"github.com/lightningnetwork/lnd/peernotifier"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// ChannelEventStore maintains a set of event logs for the node's channels.
+// It is intended to provide insight into the performance of channels.
+type ChannelEventStore struct {
+	cfg *Config
+
+	// channels maps short channel IDs to event logs.
+	channels map[uint64]*chanEventLog
+
+	quit chan struct{}
+
+	sync.Mutex
+
+	wg sync.WaitGroup
+}
+
+// Config provides the event store with functions required to monitor channel
+// activity. All elements of the config must be non-nil for the event store to
+// operate.
+type Config struct {
+	// Channel events provides a subscription client which provides a
+	// stream of channel events.
+	ChannelEvents func() (*subscribe.Client, error)
+
+	// PeerEvents provides a subscription client which provides a stream of
+	// peer online/offline events.
+	PeerEvents func() (*subscribe.Client, error)
+
+	// PeerOnline return true if a peer is online, and false if it is offline
+	// or unknown to the server.
+	PeerOnline func(peer [33]byte) bool
+
+	// Get open channels provides a list of existing open channels
+	// which is used to populate the ChannelEventStore with a set of channels
+	// on startup.
+	GetOpenChannels func() ([]*channeldb.OpenChannel, error)
+}
+
+// NewChannelEventStore initializes an event store with the config provided. Note
+// that this function does not start the monitoring process, Start() must be
+// called.
+func NewChannelEventStore(config *Config) *ChannelEventStore {
+	store := &ChannelEventStore{
+		cfg:      config,
+		channels: make(map[uint64]*chanEventLog),
+		quit:     make(chan struct{}),
+	}
+
+	return store
+}
+
+// Start adds all existing open channels to the event store and starts monitoring
+// for channel opened/closed events so that it can add and remove channels as
+// appropriate.
+func (c *ChannelEventStore) Start() error {
+	// Add the existing set of channels to the event store. This is required
+	// because channel events will not be triggered for channels that exist
+	// at startup time.
+	channels, err := c.cfg.GetOpenChannels()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Adding %v channels to event store", len(channels))
+
+	for _, ch := range channels {
+		var peerKey [33]byte
+		copy(peerKey[:], ch.IdentityPub.SerializeCompressed())
+
+		// Add existing channels to the channel store with an initial peer online
+		// or offline event. Channel opened events are not added for these
+		// channels because they are already open.
```

Comment:
doesn't the open time need to be recorded?

Reply:
Open time is currently just taken as the lowest timestamp in the event log, but I'm going to save open/close in the eventlog itself as you suggested so this will change.